extra/interface/ap/src/main/java/org/spongepowered/configurate/interfaces/processor/ConfigImplementationGenerator.java

```diff
@@ -0,0 +1,169 @@
+class ConfigImplementationGenerator {
+
+    private final ConfigImplementationGeneratorProcessor processor;
+    private final TypeElement source;
+
+    ConfigImplementationGenerator(
+        final ConfigImplementationGeneratorProcessor processor,
+        final TypeElement configInterfaceType
+    ) {
+        this.processor = processor;
+        this.source = configInterfaceType;
+    }
+
+    public TypeSpec.Builder generate() {
+        final ClassName className = ClassName.get(this.source);
+
+        this.processor.info("Generating implementation for %s", this.source);
+
+        final TypeSpec.Builder spec = TypeSpec
+            .classBuilder(className.simpleName() + "Impl")
+            .addSuperinterface(className)
+            .addModifiers(Modifier.FINAL)
+            .addAnnotation(ConfigSerializable.class)
+            .addJavadoc("Automatically generated implementation of the config");
+
+        final TypeSpecBuilderTracker tracker = new TypeSpecBuilderTracker();
+        gatherElementSpec(tracker, this.source);
+        tracker.writeTo(spec);
+
+        final String qualifiedName = className.reflectionName();
+        final String qualifiedImplName = qualifiedName.replace("$", "Impl$") + "Impl";
+        this.processor.generatedClasses().put(qualifiedName, qualifiedImplName);
+        this.processor.info("Generated implementation for %s", this.source);
+
+        return spec;
+    }
+
+    private void gatherElementSpec(
+        final TypeSpecBuilderTracker spec,
+        final TypeElement type
+    ) {
+        // first handle own elements
+
+        for (final Element enclosedElement : type.getEnclosedElements()) {
+            final ElementKind kind = enclosedElement.getKind();
+
+            if (kind == ElementKind.INTERFACE && hasAnnotation(enclosedElement, ConfigSerializable.class)) {
+                spec.add(
+                    enclosedElement.getSimpleName().toString(),
+                    new ConfigImplementationGenerator(this.processor, (TypeElement) enclosedElement)
+                        .generate()
+                        .addModifiers(Modifier.STATIC)
+                );
+                continue;
+            }
+
+            if (kind != ElementKind.METHOD) {
+                continue;
+            }
+
+            final ExecutableElement element = (ExecutableElement) enclosedElement;
+
+            if (hasAnnotation(element, PostProcess.class)) {
+                // A postprocess annotated method is not a config node
+                continue;
+            }
+
+            final boolean excluded = hasAnnotation(element, Exclude.class);
+            if (element.isDefault()) {
+                if (excluded) {
+                    // no need to handle them
+                    continue;
+                }
+                this.processor.info("Overriding implementation for %s as it's not excluded", element);
+            } else if (excluded) {
+                throw new IllegalStateException(String.format(
```

Comment:
Do not throw exceptions in APs, use the `Messager` to write out compile errors associated with the appropriate element. You do not want to crash the compiler, you want to give messages that an IDE or build tool can show to the end user -- and then keep on going, just make sure you don't generate classes for invalid input.